tests/app/views/test_agreements.py

```diff
@@ -337,6 +459,192 @@ def test_400_if_signed_agreement_details_contains_empty_strings_as_values(self,
+            }
+
+    def test_signing_framework_agreement_produces_audit_event(self, user_role_supplier, supplier_framework):
+        agreement_id = self.create_agreement(
+            supplier_framework,
+            signed_agreement_details={'signerName': 'name', 'signerRole': 'role'},
+            signed_agreement_path='/example.pdf'
+        )
+        res = self.sign_agreement(agreement_id, {'signedAgreementDetails': {'uploaderUserId': 1}})
+        assert res.status_code == 200
+
+        with self.app.app_context():
+            agreement = FrameworkAgreement.query.filter(
+                FrameworkAgreement.id == agreement_id
+            ).first()
+
+            audit = AuditEvent.query.filter(
+                AuditEvent.object == agreement
+            ).first()
+
+            assert audit.type == "sign_agreement"
+            assert audit.user == "interested@example.com"
+            assert audit.data == {
+                'supplierId': supplier_framework['supplierId'],
+                'frameworkSlug': supplier_framework['frameworkSlug'],
+                'update': {'signedAgreementDetails': {'uploaderUserId': 1}}
+            }
+
+    def test_can_resign_framework_agreement(self, user_role_supplier, supplier_framework):
+        agreement_id = self.create_agreement(
+            supplier_framework,
+            signed_agreement_details={
+                'signerName': 'name',
+                'signerRole': 'role',
+                'uploaderUserId': 2,
+                'frameworkAgreementVersion': 'v1.0'
+            },
+            signed_agreement_path='/example.pdf',
+            signed_agreement_returned_at=datetime.utcnow()
+        )
+        with freeze_time('2016-12-12'):
+            res = self.sign_agreement(agreement_id, {'signedAgreementDetails': {'uploaderUserId': 1}})
+            assert res.status_code == 200
+
+            data = json.loads(res.get_data(as_text=True))
+            assert data['agreement'] == {
+                'id': agreement_id,
+                'supplierId': supplier_framework['supplierId'],
+                'frameworkSlug': supplier_framework['frameworkSlug'],
+                'signedAgreementPath': '/example.pdf',
+                'signedAgreementDetails': {
+                    'signerName': 'name',
+                    'signerRole': 'role',
+                    'uploaderUserId': 1,
+                    'frameworkAgreementVersion': 'v1.0'
+                },
+                'signedAgreementReturnedAt': '2016-12-12T00:00:00.000000Z'
+            }
+
+    def test_can_not_sign_framework_agreement_that_has_no_signer_name(self, supplier_framework):
+        agreement_id = self.create_agreement(
+            supplier_framework,
+            signed_agreement_details={'signerRole': 'role'},
+            signed_agreement_path='/example.pdf'
+        )
+        res = self.sign_agreement(agreement_id, {'signedAgreementDetails': {'uploaderUserId': 20}})
+        assert res.status_code == 400
```

Comment:
Can we test the failure messages on these 400 tests? There are lots of ways we might get a 400 so it's nice to be sure it's the right failure that we expect.